fastly-edge-assignments/src/handlers/assignments.rs

```diff
@@ -0,0 +1,241 @@
+use eppo_core::configuration_store::ConfigurationStore;
+use eppo_core::eval::{Evaluator, EvaluatorConfig};
+use eppo_core::ufc::UniversalFlagConfig;
+use eppo_core::{Attributes, Configuration, SdkMetadata};
+use fastly::http::StatusCode;
+use fastly::kv_store::KVStoreError;
+use fastly::{Error, KVStore, Request, Response};
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+#[derive(Debug, Deserialize)]
+struct RequestBody {
+    subject_key: String,
+    subject_attributes: Arc<Attributes>,
+    // TODO: Add bandit actions
+    // #[serde(rename = "banditActions")]
+    // #[serde(skip_serializing_if = "Option::is_none")]
+    // bandit_actions: Option<HashMap<String, serde_json::Value>>,
+}
+
+// Response
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "UPPERCASE")]
+enum AssignmentFormat {
+    Precomputed,
+}
+
+#[derive(Debug, Serialize)]
+struct Environment {
+    name: String,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct FlagAssignment {
+    allocation_key: String,
+    variation_key: String,
+    variation_type: String,
+    variation_value: serde_json::Value,
+    extra_logging: HashMap<String, serde_json::Value>,
+    do_log: bool,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct AssignmentsResponse {
+    created_at: i64,
+    format: AssignmentFormat,
+    environment: Environment,
+    flags: HashMap<String, FlagAssignment>,
+}
+
+const KV_STORE_NAME: &str = "edge-assignment-kv-store";
+const SDK_KEY_QUERY_PARAM: &str = "apiKey"; // For legacy reasons this is named `apiKey`
+
+const SDK_NAME: &str = "fastly-edge-assignments";
+const SDK_VERSION: &str = "0.1.0";
+
+fn kv_store_key(token_hash: &str) -> String {
+    format!("ufc-by-sdk-key-token-hash-{}", token_hash)
+}
+
+fn token_hash(sdk_key: String) -> String {
+    let mut hasher = Sha256::new();
+    hasher.update(sdk_key.as_bytes());
+    base64_url::encode(&hasher.finalize())
+}
+
+pub fn handle_assignments(mut req: Request) -> Result<Response, Error> {
```

Comment:
minor: this function has potential to be split into multiple helper functions, which is currently hindered by the fact that error responses are constructed inline. If we have an explicit error type, that would make splitting much easier and the main body more pleasant and straightforward

Reply:
Yea I was thinking about this too - I certainly want to split out functions where reasonable but only if we think there are multiple consumers. Do you have some suggestions?

Reply:
Ok I see what you mean - ideally we return explicit errors and this handle just is responsible to rendering those into various http codes and messages?